src/OmniSharp.Roslyn/ProjectEventForwarder.cs

```diff
@@ -28,55 +30,55 @@ public class ProjectEventForwarder
             _projectSystems = projectSystems;
             _workspace = workspace;
             _emitter = emitter;
+            _eventLocks = new ConcurrentDictionary<SimpleWorkspaceEvent, AsyncLock>();
         }
 
         public void Initialize()
         {
             _workspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        private void OnWorkspaceChanged(object source, WorkspaceChangeEventArgs args)
+        private async void OnWorkspaceChanged(object source, WorkspaceChangeEventArgs args)
         {
-            SimpleWorkspaceEvent e = null;
+            SimpleWorkspaceEvent workspaceEvent = null;
 
             switch (args.Kind)
             {
                 case WorkspaceChangeKind.ProjectAdded:
-                    e = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectAdded);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectAdded);
                     break;
                 case WorkspaceChangeKind.ProjectChanged:
                 case WorkspaceChangeKind.ProjectReloaded:
-                    e = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectChanged);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectChanged);
                     break;
                 case WorkspaceChangeKind.ProjectRemoved:
-                    e = new SimpleWorkspaceEvent(args.OldSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectRemoved);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.OldSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectRemoved);
                     break;
+                default:
+                    return;
             }
 
-            if (e != null)
+            var added = false;
+            using (await _eventLocks.GetOrAdd(workspaceEvent, _ => { added = true; return new AsyncLock(); }).LockAsync())
```

Comment:
I think what this code is trying to say is:
* For a given event type, attempt to add a lock over that workspace type
* If my valueCreator is called, I am the one who added the lock
* Because of that I can lock my lock and do my processing

However, per the remarks [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netcore-3.0):

>  However, the valueFactory delegate is called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, GetOrAdd is not atomic with regards to all other operations on the ConcurrentDictionary<TKey,TValue> class.

Since a key/value can be inserted by another thread while valueFactory is generating a value, you cannot trust that just because valueFactory executed, its produced value will be inserted into the dictionary and returned. 

So I don't think this is going to have the desired effect.

Reply:
That's where something like `BlockingCollection` might work better in this scenario, we're just locking because `HashSet` is not thread safe, but in reality we want the data to push out asap, the delay doesn't need to be there for sure.

`OnWorkspaceChanged` adds an item to the queue synchronously.  Then we spin up a dedicated thread that just sits and does `TryTake(out var item, TimeSpan.Infinity)` and just writes out to the emitter.  This lets us queue these operations up, and dequeue them as quickly as possible.

Another option is to use `System.Reactive` and buffer the requests so that we pipe out all events every <500 ms (or whatever)>.

The blocking collection works well for LSP, so it should probably work okay here.